apps/prairielearn/src/lib/assessment.ts

```diff
@@ -182,22 +182,29 @@ export async function gradeAssessmentInstanceAsync(
   // grading transaction has been accepted; collect those job ids here.
   const externalGradingJobIds: string[] = [];
 
-  if (requireOpen) {
-    await sqldb.callAsync('assessment_instances_ensure_open', [assessment_instance_id]);
+  if (requireOpen || close) {
+    await sqldb.runInTransactionAsync(async () => {
+      await sqldb.callAsync('assessment_instances_ensure_open', [assessment_instance_id]);
+
+      if (close) {
+        // If we're supposed to close the assessment, do it *before* we
+        // we start grading. This avoids a race condition where the student
+        // makes an additional submission while grading is already in progress.
+        await sqldb.queryAsync(sql.close_assessment_instance, {
```

Comment:
The `assessment_instances_close` had an explicit `assessment_instances_lock()` call that would guard a) the check if the instance is open and b) the close. I don't think the new code has exactly the same behavior; should this transaction lock the assessment instance at the beginning?

Reply:
Good catch. For some reason I had erroneously considered that `assessment_instances_ensure_open` also locked the assessment_instance record. I'll review this. Would you be ok if `...ensure_open` always locks the instance, provided I rearrange the other uses of this function (in essence the creation of variants and the creation of new submissions) to avoid deadlocks (by ensuring they're always locked in the same order)?

Reply:
I suppose I'd lean towards just adding another sproc call inside the transaction instead of changing the behavior of an existing sproc, especially in a way that would make it inconsistent with how the other `*_ensure_open` sprocs work.

Reply:
Done in ebe452e83.